xtm/xtm_scsp_queue.h

```diff
@@ -0,0 +1,173 @@
+	 *                   for fast calculation of read/write position, using
+	 *                   '& (size - 1' operation.
+	 * @retval 0 if success, otherwise return -1.
+	 */
+	int
+	create(unsigned size) {
+		/* Ensure size is power of 2 */
+		if (size & (size - 1) || size <= 1)
+			return -1;
+
+		write = 0;
+		read = 0;
+		len_minus_1 = size - 1;
+		return 0;
+	}
+	/**
+	 * Add num elements into queue
+	 * @param[in] data - array of elements to add
+	 * @param[in] num - count of elements to add
+	 * @retval number of elements actually written.
+	 */
+	unsigned
+	put(T *data, unsigned num) {
+		unsigned i;
+		unsigned queue_write = write;
+		unsigned new_write = queue_write;
+		unsigned queue_read = __atomic_load_n(&read, __ATOMIC_ACQUIRE);
+
+		for (i = 0; i < num; i++) {
+			new_write = (new_write + 1) & len_minus_1;
+			if (new_write == queue_read)
+				break;
+			buffer[queue_write] = data[i];
+			queue_write = new_write;
+		}
+		__atomic_store_n(&write, queue_write, __ATOMIC_RELEASE);
+		return i;
+	}
+	/**
+ 	 * Get num of available elements in the queue
+	 * @retval num of available elements in the queue
+	 */
+	unsigned
+	free_count(void) {
+		unsigned queue_write = __atomic_load_n(&write, __ATOMIC_ACQUIRE);
+		unsigned queue_read = __atomic_load_n(&read, __ATOMIC_ACQUIRE);
+		return (queue_read - queue_write - 1) & len_minus_1;
+	}
+	/**
+	 * Get num of elements in the queue
+	 * @retval return num of elements in the queue
+	 */
+	unsigned
+	count(void) {
+		unsigned queue_write = __atomic_load_n(&write, __ATOMIC_ACQUIRE);
+		unsigned queue_read = __atomic_load_n(&read, __ATOMIC_ACQUIRE);
+		return (len_minus_1 + 1 + queue_write - queue_read) & len_minus_1;
+	}
+private:
+	/** Next position to be written */
+	unsigned write;
+	/** Next position to be read */
+	unsigned read;
+	/** Circular buffer length */
+	unsigned len_minus_1;
+	/** Buffer contains objects */
+	T buffer[];
+};
+
+/**
+ * Read iterator for single consumer, single producer queue.
+ * Template parameter must be same, as in queue to iterate.
+ */
+template <class T>
+struct xtm_scsp_queue_read_iterator {
+	/**
+	 * Create new queue read iterator
+	 * @param[in] queue - queue to iterate
+	 */
+	void
+	begin(struct xtm_scsp_queue<T> *q) {
+		queue = q;
+		read_pos = queue->read;
+		end_of_read = __atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+		is_queue_full = (((read_pos - end_of_read - 1) &
+			queue->len_minus_1) == 0);
+	}
+	/**
+	 * Read next element from queue.
+	 * @retval next element from queue.
+	 */
+	const T*
+	read(void) {
+		if (read_pos != end_of_read) {
+			const T *rc = &queue->buffer[read_pos];
+			read_pos = (read_pos + 1) & queue->len_minus_1;
+			return rc;
+		}
+		return nullptr;
+	}
+	/**
+	 * Store new read index to the queue.
+	 */
+	void
+	end(void) {
+		__atomic_store_n(&queue->read, read_pos, __ATOMIC_RELEASE);
+	}
+	/**
+	 * Return flag, which indicates, that queue
+	 * was full, when iter created.
+	 */
+	bool
+	is_queue_was_full(void) {
```

Comment:
Please, just `was_queue_full()`.

Reply:
Remove this function at all, because of API changes